templates/cluster-template-flatcar.yaml

```diff
@@ -0,0 +1,265 @@
+          mountPath: /etc/kubernetes/azure.json
+          name: cloud-config
+          readOnly: true
+      etcd:
+        local:
+          dataDir: /var/lib/etcddisk/etcd
+          extraArgs:
+            quota-backend-bytes: "8589934592"
+    diskSetup:
+      filesystems: []
+      partitions: []
+    files:
+    - contentFrom:
+        secret:
+          key: control-plane-azure.json
+          name: ${CLUSTER_NAME}-control-plane-azure-json
+      owner: root:root
+      path: /etc/kubernetes/azure.json
+      permissions: "0644"
+    format: ignition
+    ignition:
+      containerLinuxConfig:
+        additionalConfig: |
+          systemd:
+            units:
+            - name: kubeadm.service
+              dropins:
+              - name: 10-flatcar.conf
+                contents: |
+                  [Unit]
+                  After=oem-cloudinit.service
+          # Workaround for https://github.com/kubernetes-sigs/cluster-api/issues/7679.
+          storage:
+            disks:
+            - device: /dev/disk/azure/scsi1/lun0
+              partitions:
+              - number: 1
+            filesystems:
+            - name: etcd_disk
+              mount:
+                device: /dev/disk/azure/scsi1/lun0
+                format: ext4
+                label: etcd_disk
+                options:
+                - -E
+                - lazy_itable_init=1,lazy_journal_init=1
+    initConfiguration:
+      nodeRegistration:
+        kubeletExtraArgs:
+          azure-container-registry-config: /etc/kubernetes/azure.json
+          cloud-config: /etc/kubernetes/azure.json
+          cloud-provider: azure
+        name: '@@HOSTNAME@@'
+    joinConfiguration:
+      nodeRegistration:
+        kubeletExtraArgs:
+          azure-container-registry-config: /etc/kubernetes/azure.json
+          cloud-config: /etc/kubernetes/azure.json
+          cloud-provider: azure
+        name: '@@HOSTNAME@@'
+    mounts:
+    - - etcd_disk
+      - /var/lib/etcddisk
+    postKubeadmCommands: []
+    preKubeadmCommands:
+    - sed -i "s/@@HOSTNAME@@/$(curl -s -H Metadata:true --noproxy '*' 'http://169.254.169.254/metadata/instance?api-version=2020-09-01'
+      | jq -r .compute.name)/g" /etc/kubeadm.yml
+  machineTemplate:
+    infrastructureRef:
+      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+      kind: AzureMachineTemplate
+      name: ${CLUSTER_NAME}-control-plane
+  replicas: ${CONTROL_PLANE_MACHINE_COUNT}
+  version: ${KUBERNETES_VERSION}
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureMachineTemplate
+metadata:
+  name: ${CLUSTER_NAME}-control-plane
+  namespace: default
+spec:
+  template:
+    spec:
+      dataDisks:
+      - diskSizeGB: 256
+        lun: 0
+        nameSuffix: etcddisk
+      image:
+        computeGallery:
+          gallery: flatcar4capi-742ef0cb-dcaa-4ecb-9cb0-bfd2e43dccc0
+          name: flatcar-stable-amd64-capi-${KUBERNETES_VERSION}
+          version: 3374.2.0
```

Comment:
why are we hardcoding a specific version? is the version guaranteed to be the same for every image / k8s version going forward?

Reply:
So, for end users, we recommend explicitly selecting the version of Flatcar to use, as changing the version in machine deployment should trigger machines replacement and perform a cluster update.

For CI here, my assumption was that when KUBERNETES_VERSION used by CI changes, this will probably change as well, as we likely want to build latest K8s version images with latest Flatcar stable, but I _think_ we could patch it to just use `latest`, although we do plan to exclude all CAPI versions from latest so they must be selected explicitly.

Reply:
An alternative would be to introduce `FLATCAR_VERSION` and maybe `FLATCAR_CHANNEL` variables?

Reply:
I ended up introducing `FLATCAR_VERSION` variable, which in e2e tests is automatically discovered for available K8s version. I'm omitting `FLATCAR_CHANNEL` for the time being, as we plan to only publish `stable` images for the time being to avoid large number of images being published.

I'll mark it as resolved, as I think it's also related to version discovery discussion we have in another thread.